File: LeetCode_POTD/June_2026/June_12.py
"""
As there are multiple different queries we have to optimize the way to find distance between 2 nodes of query -> Lowest Common Ancestor(Binary Lifting)( O(logn) per query, processing time O(nlogn) )
To prevent stack overflow for skewed tree cases- we prefer BFS over DFS
1. Building Tree: (O(nlogn))
    Make arraylist for adjacent nodes of each node
    Using Queue traverse BFS starting by Node 1 with mentioning depth of node in array depth
    Define LOG for maximum size of ancestors going up on base 2(not mentioning each node)--> up[n+1][LOG]
    up[node'val][ancestor no.] (If ancestor no. = 0 -> 1 step up, 1 -> 2 step up, 2 -> 4 step up ...) (up[node][j] =up[up[node][j - 1]][j - 1], j=ancestor no.)
2. Resolving Query: O(logn)
    Pass (dist-1) between node in solvex function to get answer(2^(dist-1)%1_000_000_007)
    How to get dist:
        Firstly find differnce in depth, then make both at same depth(min of both node's depth)-> To make depth equal use binary system to update node
        Then from the top(root), check with decreasing ancestor no. to react lowest commmon ancestror(if different ancestor then , safe to move upward)

"""

class Solution:
    def assignEdgeWeights(self, edges: List[List[int]], queries: List[List[int]]) -> List[int]:
        n=len(edges)+1
        LOG=1
        depth=[0]*(n+1)
        while 1<<LOG<=n: LOG+=1
        up=[[0 for _ in range(LOG)] for _ in range(n+1)]
        adj=[[] for _ in range(n+1)]
        for i,j in edges:
            adj[i].append(j)
            adj[j].append(i)
        queue=deque()
        queue.append(1)
        depth[1]=0
        up[1][0]=1
        vis=[False]*(n+1)
        vis[1]=True
        while queue:
            i=queue.popleft()
            for j in adj[i]:
                if not vis[j]:
                    queue.append(j)
                    depth[j]=depth[i]+1
                    up[j][0]=i
                    vis[j]=True
        for i in range(1,LOG):
            for j in range(1,n+1):
                up[j][i]=up[up[j][i-1]][i-1]

        MOD = 1000000007
        pow2 = [1] * n
        for i in range(1, n):
            pow2[i] = (pow2[i - 1] * 2) % MOD

        ans=[]
        for i,j in queries:
            dist=depth[i]+depth[j]
            if depth[i]<depth[j]:
                i,j=j,i
            diff=depth[i]-depth[j]
            k=0
            while 1<<k <= diff:
                if diff&(1<<k)!=0:
                    i=up[i][k]
                k+=1
            if i==j: dist-=2*depth[i]
            else:
                for o in range(LOG-1,-1,-1):
                    if up[i][o]!=up[j][o]:
                        i=up[i][o]
                        j=up[j][o]
                dist-=2*depth[up[i][0]]
            if dist <= 0:
                ans.append(0)
            else:
                ans.append(pow2[dist - 1])
        return ans

File: LeetCode_POTD/June_2026/June_12.java
package Leetcode_POTD.June_2026;

/*
As there are multiple different queries we have to optimize the way to find distance between 2 nodes of query -> Lowest Common Ancestor(Binary Lifting)( O(logn) per query, processing time O(nlogn) )
To prevent stack overflow for skewed tree cases- we prefer BFS over DFS
1. Building Tree: (O(nlogn))
    Make arraylist for adjacent nodes of each node
    Using Queue traverse BFS starting by Node 1 with mentioning depth of node in array depth
    Define LOG for maximum size of ancestors going up on base 2(not mentioning each node)--> up[n+1][LOG]
    up[node'val][ancestor no.] (If ancestor no. = 0 -> 1 step up, 1 -> 2 step up, 2 -> 4 step up ...) (up[node][j] =up[up[node][j - 1]][j - 1], j=ancestor no.)
2. Resolving Query: O(logn)
    Pass (dist-1) between node in solvex function to get answer(2^(dist-1)%1_000_000_007)
    How to get dist:
        Firstly find differnce in depth, then make both at same depth(min of both node's depth)-> To make depth equal use binary system to update node
        Then from the top(root), check with decreasing ancestor no. to react lowest commmon ancestror(if different ancestor then , safe to move upward)
*/

import java.util.ArrayDeque;
import java.util.ArrayList;

public class June_12 {
    private int[][] up;
    private int[] depth;
    private int LOG;

    public int[] assignEdgeWeights(int[][] edges, int[][] queries) {
        int n = edges.length + 1;
        LOG = 1;
        while ((1 << LOG) <= n) LOG++;
        depth = new int[n + 1];
        up = new int[n + 1][LOG];
        buildTreeBFS(edges, n);

        int[] ans = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int u = queries[i][0];
            int v = queries[i][1];
            int dist = distance(u, v);
            ans[i] = solvex(2, dist - 1, 1_000_000_007);
        }

        return ans;
    }
    private void buildTreeBFS(int[][] edges, int n) {
        ArrayList<Integer>[] adj = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            adj[i] = new ArrayList<>();
        }
        for (int[] e : edges) {
            adj[e[0]].add(e[1]);
            adj[e[1]].add(e[0]);
        }
        boolean[] vis = new boolean[n + 1];
        ArrayDeque<Integer> q = new ArrayDeque<>();
        q.offer(1);
        vis[1] = true;
        depth[1] = 0;
        up[1][0] = 1;
        while (!q.isEmpty()) {
            int cur = q.poll();
            for (int nxt : adj[cur]) {
                if (!vis[nxt]) {
                    vis[nxt] = true;
                    depth[nxt] = depth[cur] + 1;
                    up[nxt][0] = cur;
                    q.offer(nxt);
                }
            }
        }
        for (int j = 1; j < LOG; j++) {
            for (int node = 1; node <= n; node++) {
                up[node][j] =up[up[node][j - 1]][j - 1];
            }
        }
    }
    private int lca(int a, int b) {
        if (depth[a] < depth[b]) {
            int temp = a;
            a = b;
            b = temp;
        }
        int diff = depth[a] - depth[b];
        for (int j = LOG - 1; j >= 0; j--) {
            if ((diff & (1 << j)) != 0) {
                a = up[a][j];
            }
        }
        if (a == b) {
            return a;
        }
        for (int j = LOG - 1; j >= 0; j--) {
            if (up[a][j] != up[b][j]){
                a = up[a][j];
                b = up[b][j];
            }
        }
        return up[a][0];
    }
    private int distance(int u, int v) {
        int lca = lca(u, v);
        return depth[u]+ depth[v]- 2 * depth[lca];
    }
    private int solvex(long base, long exp, int mod) {
        if (exp < 0) return 0;
        long res = 1;
        while (exp > 0) {
            if ((exp & 1) == 1) {
                res = (res * base) % mod;
            }
            base = (base * base) % mod;
            exp >>= 1;
        }
        return (int) res;
    }
}
